ecs-init/cache/dependencies.go

```diff
@@ -26,15 +26,98 @@ import (
 	"os"
 	"path/filepath"
 
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/s3"
 	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	log "github.com/cihub/seelog"
+	"github.com/pkg/errors"
 )
 
-// s3Downloader captures the only method used from the s3 client
-type s3Downloader interface {
+// s3API captures the only method used from the s3 package
+type s3API interface {
 	Download(w io.WriterAt, input *s3.GetObjectInput, options ...func(*s3manager.Downloader)) (n int64, err error)
 }
 
+// s3BucketDownloader wraps a bucket together with a downloader that can download from it
+type s3BucketDownloader struct {
+	bucket string
+	region string
+	client s3API
+}
+
+func newS3BucketDownloader(region, bucketName string) (*s3BucketDownloader, error) {
+	session, err := session.NewSession(&aws.Config{
+		Credentials: credentials.AnonymousCredentials,
+		Region:      aws.String(region),
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to initialize downloader in region %s", region)
+	}
+
+	s3BucketDownloader := &s3BucketDownloader{
+		client: s3manager.NewDownloader(session),
+		bucket: bucketName,
+		region: region,
+	}
+
+	return s3BucketDownloader, nil
+}
+
+func (bd *s3BucketDownloader) download(fileName, cacheDir string, fs fileSystem) (string, error) {
+	file, err := fs.TempFile(cacheDir, fileName)
+	if err != nil {
+		return "", errors.Wrap(err, "could not create local file during download")
+	}
+
+	defer func() { // make sure we also handle possible error from f.Close()
+		cerr := file.Close()
+		if err == nil { // if no error return from download, captures error from f.Close()
+			err = cerr
+		}
+	}()
+
+	_, err = bd.client.Download(file, &s3.GetObjectInput{
+		Bucket: aws.String(bd.bucket),
+		Key:    aws.String(fileName),
+	})
+
+	return file.Name(), err
+}
+
+type s3DownloaderAPI interface {
+	addBucketDownloader(bucketDownloader *s3BucketDownloader)
+	downloadFile(fileName string) (string, error)
+}
+
+type s3Downloader struct {
+	bucketDownloaders []*s3BucketDownloader
+	fs                fileSystem
+	cacheDir          string
+}
+
+func (d *s3Downloader) addBucketDownloader(bucketDownloader *s3BucketDownloader) {
+	d.bucketDownloaders = append(d.bucketDownloaders, bucketDownloader)
+}
+
+func (d *s3Downloader) downloadFile(fileName string) (string, error) {
+	for _, bucketDownloader := range d.bucketDownloaders {
```

Comment:
I don't like the idea to implicitly retry with different downloaders. A better approach could be explicitly call the download function of the different downloader. This can be done in a separate pr if you want, so that this is not blocked.
```
downloadWithPartitionBucket()
downloadWithRegionalBucket()
```

Reply:
i'm doing it implicitly due to other comments.. i guess we can discuss this offline and change it in another pr later